### PR TITLE
[pkg/ottl] add time converter function

### LINKED ISSUE #22007: [pkg/ottl] OTTL Converter Function that is able to create a new `time.Time` type from a string.
Create an OTTL Converter Function(s), probably called `Time`, that is able to create a new `time.Time` type from a string.  Something similar to https://pkg.go.dev/time#Parse or https://github.com/knz/strtime or https://github.com/itchyny/timefmt-go can be used.

The function should take 2 parameters, a string (represented as a `StringGetter`) that represents the time and a string (represented as a `StringGetter`) that represents the format that the time should be parse.  We should also support standard names for common formats in the second parameter, like `RFC3339` or `ISO 8601` that the function would know how to interpret.

### PR DESCRIPTION
**Pull request by fchikwekwe**
Description: Added a converter function that takes a time string and a format string and converts to a golang time.Time object. 

Link to tracking Issue: #22007 

Testing: Unit tests to check proper conversion from `stringGetter` to golang `time.Time`

Documentation:

### CODE REVIEW

File: pkg/ottl/ottlfuncs/func_to_time.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type ToTimeArguments[K any] struct {
+	Time   ottl.StringGetter[K] `ottlarg:"0"`
+	Format ottl.StringGetter[K] `ottlarg:"1"`
+}
+
+func NewToTimeFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("toTime", &ToTimeArguments[K]{}, createToTimeFunction[K])
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_to_time.go:20]: OTTL requires that converter functions start with a capital letter.  To be consistent with our other type-creating convertes such as `Int`, `SpanID`, and `TraceID`, let's call this function `Time` and update the file's function names accordingly.

File: pkg/ottl/ottlfuncs/func_to_time_test.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+				Getter: func(ctx context.Context, tCtx interface{}) (interface{}, error) {
+					return "11/11/11", nil
+				},
+			},
+			format: &ottl.StandardTypeGetter[interface{}, string]{
+				Getter: func(ctx context.Context, tCtx interface{}) (interface{}, error) {
+					return "", nil
+				},
+			},
+			expectedError: "extra text: \"11/11/11\"",
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_to_time_test.go:251]: I like that `time.Parse` can handle this as an error, but it's not as helpful as I was hoping.  Lets add a specific case in the function before the switch statement checking for `""` in the format like we do for the input.

File: pkg/ottl/ottlfuncs/func_to_time_test.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+			assert.NoError(t, err)
+			result, err := exprFunc(nil, nil)
+			assert.NoError(t, err)
+			assert.Equal(t, tt.expected, result)
+		})
+	}
+}
+
+func Test_toTimeError(t *testing.T) {
+	tests := []struct {
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_to_time_test.go:219]: Please add a test case for empty string input.

File: pkg/ottl/ottlfuncs/func_to_time_test.go [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+		{
+			name: "simple short form",
+			time: &ottl.StandardTypeGetter[interface{}, string]{
+				Getter: func(ctx context.Context, tCtx interface{}) (interface{}, error) {
+					return "2023-04-12", nil
+				},
+			},
+			format: &ottl.StandardTypeGetter[interface{}, string]{
+				Getter: func(ctx context.Context, tCtx interface{}) (interface{}, error) {
+					return "2006-01-02", nil
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_to_time_test.go:33]: I think it is because I've used other formats more, but after seeing these test cases I find Go's format logic very jarring.  I was definitely expecting the format to look like `"%Y-%m-%d"`.  Since many collector users are not Go users, I think if we can find a solid library that lets us use strptime format that will be preferred.

**Review by TylerHelmuth — COMMENTED** [review]
Great stuff! Please update the ottlfuncs/README with a new entry for this function.

**Review by TylerHelmuth — COMMENTED** [review]
LGTM. Please add a changelog entry by running `chlog-new` in the root directory of the repo.  Also update `pkg/ottl/go.mod` with the new package dependency.  Once you've updated the go.mod file run `make gotidy` in the root directory of the repo.

File: pkg/ottl/ottlfuncs/func_time_test.go [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+func Test_Time(t *testing.T) {
+	tests := []struct {
+		name     string
+		time     ottl.StringGetter[interface{}]
+		format   ottl.StringGetter[interface{}]
+		expected interface{}
+	}{
+		{
+			name: "simple short form",
+			time: &ottl.StandardTypeGetter[interface{}, string]{
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_time_test.go:26]: `StandardTypeGetter` should be replaced with `StringGetter`.

Comment (fchikwekwe) [reply pkg/ottl/ottlfuncs/func_time_test.go]: Just pushed an update that fixes this

**Review by TylerHelmuth — DISMISSED** [review]
LGTM!

File: pkg/ottl/ottlfuncs/func_time.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_time.go:4]:
```suggestion
package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
```

**Comment by TylerHelmuth** [comment]
While investigating these tests we found inconsistencies with timezones (it's always timezones).  Comparing the time parsing to `pkg/stanza/operator/helper` `TimeParser` we found that using only the `ctimefmt` package was not enough as it did not handle timezones the same way `TimeParser does`. 

Investigating further we found that this was the magic that made `TimeParser` work as it does:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/d0fe5f310ca8be484e5c83c9c772144d322e5859/pkg/stanza/operator/helper/time.go#L181-L204

and

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/d0fe5f310ca8be484e5c83c9c772144d322e5859/pkg/stanza/operator/helper/time.go#L113-L130

Handling the time in this way is necessary for a consistent time parsing strategy in the collector.  Instead of duplicating these lines we should move the functionality out of stanza and into a shared place.  See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23232 for implementation.

File: pkg/ottl/ottlfuncs/func_time.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	"fmt"
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/timeutils"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type TimeArguments[K any] struct {
+	Time     ottl.StringGetter[K] `ottlarg:"0"`
+	Format   ottl.StringGetter[K] `ottlarg:"1"`
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_time.go:17]: Let's change `Format` to be a `string` instead of a `StringGetter`. This brings Time inline with other functions that take a format, like a regex pattern, all of which force the value to be statically defined in the config via a string. Also, having a static string format will allow us to call `GetLocation` inside `time` before the return function.  This will give use the location to use so it doesn't need to be passed in.

File: pkg/ottl/ottlfuncs/func_time.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+	if !ok {
+		return nil, fmt.Errorf("TimeFactory args must be of type *TimeArguments[K]")
+	}
+
+	return Time(args.Time, args.Format, args.Location)
+}
+
+func Time[K any](inputTime ottl.StringGetter[K], format ottl.StringGetter[K], location ottl.StringGetter[K]) (ottl.ExprFunc[K], error) {
+	return func(ctx context.Context, tCtx K) (interface{}, error) {
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_time.go:35]: Before this return statement we can to some setup:

```go
if format == "" {
  return nil, fmt.Errorf("format cannot be empty")
}
loc := timeutils.GetLocation(nil, format)
return ...
```

File: pkg/ottl/ottlfuncs/func_time.go (hunk 7 above)
Comment (fchikwekwe) [reply pkg/ottl/ottlfuncs/func_time.go]: Changed!

File: pkg/ottl/ottlfuncs/func_time.go (hunk 8 above)
Comment (fchikwekwe) [reply pkg/ottl/ottlfuncs/func_time.go:41]: Changed and some test changes to reflect this.

File: pkg/ottl/ottlfuncs/func_time_test.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+		{
+			name: "simple short form with long year and slashes",
+			time: &ottl.StandardStringGetter[interface{}]{
+				Getter: func(ctx context.Context, tCtx interface{}) (interface{}, error) {
+					return "02/12/2022", nil
+				},
+			},
+			format:   "%m/%d/%Y",
+			expected: time.Date(2022, 2, 12, 0, 0, 0, 0, time.Local),
+		},
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_time_test.go:63]: Are these 2 tests the same?

**Review by TylerHelmuth — APPROVED** [review]

**Comment by TylerHelmuth** [comment]
@evan-bradley can you review this one as well?  I'm too close to it after doing the refactor in timeutils to unblock it.

File: pkg/ottl/ottlfuncs/func_time_test.go (hunk 9 above)
Comment (fchikwekwe) [reply pkg/ottl/ottlfuncs/func_time_test.go:63]: Hm, yeah this must have happened while I was changing tests around. The duplicate is removed.

**Review by djaglowski — APPROVED** [review]

**Comment by ptimofee** [comment]
> Great stuff! Please update the ottlfuncs/README with a new entry for this function.

Not done

**Comment by TylerHelmuth** [comment]
@ptimofee you are correct, I missed that. It is now being added via https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23659
